### PR TITLE
Authtoken

### PR DESCRIPTION
**Pull request by tekrajchhetri**
This pull requests implements the following:

-Token based authentication & authorization

@puja-trivedi @aaronkanzer

### CODE REVIEW

File: core/routers/index.py [hunk 1]
```diff
@@ -10,16 +10,26 @@
...
 
 @router.get("/")
 async def root():
-    return {"message": "Hello World"}
+    return {"message": "Welcome to FastAPI skeleton"}
+
+
+@router.get("/token-check", dependencies=[Depends(require_scopes(["read"]))])
+async def token_check(user: Annotated[LoginUserIn, Depends(get_current_user)]):
+    return {"message": "token check passed sucecss"}
```
Comment (aaronkanzer) [inline core/routers/index.py:35]: typo: `success`

File: core/security.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+from fastapi.security import HTTPAuthorizationCredentials, HTTPBearer
+from fastapi.security import OAuth2PasswordBearer
+from jose import ExpiredSignatureError, JWTError, jwt
+from passlib.context import CryptContext
+
+from core.database import get_user
+
+logger = logging.getLogger(__name__)
+
+SECRET_KEY = "9b73f2a1bdd7ae163444473d29a6885ffa22ab26117068f72a5a56a74d12d1fc"
```
Comment (aaronkanzer) [inline core/security.py:33]: Secret key should not be in plain-text in commits -- please remove, invalidate, and renew

**Comment by aaronkanzer** [comment]
@tekrajchhetri Could you still look into AWS Cognito as a service? I appreciate the effort here to use custom code; however, I think maintenance, cognitive overhead, security, unit tests, etc. might warrant a managed service here.

Any reason you decided not to go the Cognito route here? I recall we might have previously discussed this in a prior PR.

**Comment by tekrajchhetri** [comment]
> @tekrajchhetri One other thing regarding design decision here (if you still decide to go with the non-Cognito route):
> 
> You are storing sessions (e.g. checking the current user) via a relational database in your code here -- querying for valid sessions is a frequent task (e.g. you'd have to authenticate the user per permissioned request, etc.)
> 
> I'd recommend instead going a key-value DB route, such as AWS DynamoDB, Redis, Memcache, etc. -- these are often the industry-standard for these types of queries.
> 
> Happy to explain further!

Thanks! I will check the key-value DB

File: core/security.py (hunk 2 above)
Comment (tekrajchhetri) [reply core/security.py:33]: yes, it should not be in the code but in env file.  I had it in a env file, will check and update it. Thanks!

Comment (tekrajchhetri) [reply core/security.py]: I have updated the code.

File: core/routers/index.py (hunk 1 above)
Comment (tekrajchhetri) [reply core/routers/index.py]: corrected!

**Comment by tekrajchhetri** [comment]
@aaronkanzer @puja-trivedi I think this can be merged as it's been long pending and the issues that @aaronkanzer raised was corrected/fixed, except the one for key-value where I decided to go with the current Postgres.

**Comment by aaronkanzer** [comment]
> @aaronkanzer @puja-trivedi I think this can be merged as it's been long pending and the issues that @aaronkanzer raised was corrected/fixed, except the one for key-value where I decided to go with the current Postgres.

Sounds good to me -- just out of curiosity, why did you decide to stay with the current Postgres approach rather than a key-value approach for session management?

**Comment by tekrajchhetri** [comment]
> Sounds good to me -- just out of curiosity, why did you decide to stay with the current Postgres approach rather than a key-value approach for session management?

@aaronkanzer While I understand the advantage of key-value approach, currently for us, it wouldn't make much difference even with Postgres and we can switch it later to key-value as needed. The other reason was as it's working, I can focus on more important, i.e., BrainKB's core task like knowledge extraction. 

Also, could you approve the merge?

**Review by aaronkanzer — APPROVED** [review]

**Comment by aaronkanzer** [comment]
> > Sounds good to me -- just out of curiosity, why did you decide to stay with the current Postgres approach rather than a key-value approach for session management?
> 
> @aaronkanzer While I understand the advantage of key-value approach, currently for us, it wouldn't make much difference even with Postgres and we can switch it later to key-value as needed. The other reason was as it's working, I can focus on more important, i.e., BrainKB's core task like knowledge extraction.
> 
> Also, could you approve the merge?

I've approved.

My advice here however is not accumulating tech debt this early on when these implementations are lower-lift (probably would take an hour at-most to provision a Redis container in Docker or route to a Dynamo Table instead of Postgres). 

Refactoring is complex and time-consuming, and based off of the goals of BrainKB, this is as much a web development and site reliability challenge as it is a planar graph traversal and construction challenge -- always happy to review nevertheless, but keep in mind software engineering works best as a marathon, not a sprint 😄

**Comment by tekrajchhetri** [comment]
sure, thank you!

**Pull request merged**